std/algorithm/iteration.d

```diff
@@ -4035,16 +4038,97 @@ if (isInputRange!R && !isInfinite!R && is(typeof(seed = seed + r.front)))
 }
 
 // Pairwise summation http://en.wikipedia.org/wiki/Pairwise_summation
-private auto sumPairwise(Result, R)(R r)
+private auto sumPairwise(F, R)(R data)
+if (isInputRange!R && !isInfinite!R)
 {
-    static assert (isFloatingPoint!Result);
-    switch (r.length)
+    import core.bitop : bsf;
+    // Works for r with at least length < 2^^(64 + log2(16)), in keeping with the use of size_t
+    // elsewhere in std.algorithm and std.range on 64 bit platforms. The 16 in log2(16) comes
+    // from the manual unrolling in sumPairWise16
+    F[64] store = void;
+    size_t idx = 0;
+
+    auto collapseStore(T)(T k)
```

Comment:
#4013 takes the approach of asserting that the length of the range will fit in `size_t`, so `long` and `ulong` will work on 32-bit systems - but only if the resultant range is not too long. So, most common cases will work just fine, but if you try and do something like `iota(long.min, long.max)`, then the assertion will fail. The downside, of course, is that when the problem does occur, it's only found at runtime, but in the vast majority of cases, you can continue to use `long` and `ulong` with `iota` and have `length` as we've had.